File: 56-merge-intervals/56-merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        if len(intervals) <= 1:
            return intervals
        ans = []
        intervals = sorted(intervals, key = lambda interval: interval[0])
        low, high = intervals[0][0], intervals[0][1]
        for interval in intervals:
            if interval[0] <= high:
                high = max(interval[1], high)
            else:
                ans.append([low, high])
                high = interval[1]
                low = interval[0]
        ans.append([low, high])
        return ans
    
#Approach: Sort the intervals by start. Traverse the array and keep updating high till there is overlap. Initialise low and high after adding to ans once there is no overlap.

File: 56-merge-intervals/56-merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        if (intervals.length <= 1) {
            return intervals;
        }
        List<int[]> mergedList = new ArrayList<>();
        Arrays.sort(intervals, Comparator.comparingInt(arr -> arr[0]));

        int low = intervals[0][0];
        int high = intervals[0][1];

        for(int[] interval : intervals) {
            if(interval[0] > high) {
                mergedList.add(new int[] {low, high});
                low = interval[0];
                high = interval[1];
            } else {
                high = Math.max(interval[1], high);
            }
        }
        mergedList.add(new int[]{low, high});

        return mergedList.toArray(new int[mergedList.size()][]);
    }
}